server/start.go

```diff
@@ -640,3 +614,364 @@ func startApp(svrCtx *Context, appCreator types.AppCreator, opts StartCmdOptions
+				text, _ := reader.ReadString('\n')
+				response := strings.TrimSpace(strings.ToLower(text))
+				if response != "y" && response != "yes" {
+					fmt.Println("Operation canceled.")
+					return nil
+				}
+			}
+
+			// Set testnet keys to be used by the application.
+			// This is done to prevent changes to existing start API.
+			serverCtx.Viper.Set(KeyIsTestnet, true)
+			serverCtx.Viper.Set(KeyNewChainID, newChainID)
+			serverCtx.Viper.Set(KeyNewOpAddr, newOperatorAddress)
+
+			err = wrapCPUProfile(serverCtx, func() error {
+				return start(serverCtx, clientCtx, testnetAppCreator, withCMT, opts)
+			})
+
+			serverCtx.Logger.Debug("received quit signal")
+			graceDuration, _ := cmd.Flags().GetDuration(FlagShutdownGrace)
+			if graceDuration > 0 {
+				serverCtx.Logger.Info("graceful shutdown start", FlagShutdownGrace, graceDuration)
+				<-time.After(graceDuration)
+				serverCtx.Logger.Info("graceful shutdown complete")
+			}
+
+			return err
+		},
+	}
+
+	addStartNodeFlags(cmd, opts)
+	cmd.Flags().String(KeyTriggerTestnetUpgrade, "", "If set (example: \"v21\"), triggers the v21 upgrade handler to run on the first block of the testnet")
+	cmd.Flags().Bool("skip-confirmation", false, "Skip the confirmation prompt")
+	return cmd
+}
+
+// testnetify modifies both state and blockStore, allowing the provided operator address and local validator key to control the network
+// that the state in the data folder represents. The chainID of the local genesis file is modified to match the provided chainID.
+func testnetify(ctx *Context, home string, testnetAppCreator types.AppCreator, db dbm.DB, traceWriter io.WriteCloser) (types.Application, error) {
+	config := ctx.Config
+
+	newChainID, ok := ctx.Viper.Get(KeyNewChainID).(string)
+	if !ok {
+		return nil, fmt.Errorf("expected string for key %s", KeyNewChainID)
+	}
+
+	genDocProvider := node.DefaultGenesisDocProviderFunc(config)
+
+	// Initialize blockStore and stateDB.
+	blockStoreDB, err := cmtcfg.DefaultDBProvider(&cmtcfg.DBContext{ID: "blockstore", Config: config})
+	if err != nil {
+		return nil, err
+	}
+	blockStore := store.NewBlockStore(blockStoreDB)
+
+	stateDB, err := cmtcfg.DefaultDBProvider(&cmtcfg.DBContext{ID: "state", Config: config})
+	if err != nil {
+		return nil, err
+	}
+
+	defer blockStore.Close()
+	defer stateDB.Close()
+
+	privValidator := pvm.LoadOrGenFilePV(config.PrivValidatorKeyFile(), config.PrivValidatorStateFile())
+	userPubKey, err := privValidator.GetPubKey()
+	if err != nil {
+		return nil, err
+	}
+	validatorAddress := userPubKey.Address()
+	if err != nil {
+		return nil, err
+	}
+
+	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
+		DiscardABCIResponses: config.Storage.DiscardABCIResponses,
+	})
+
+	state, genDoc, err := node.LoadStateFromDBOrGenesisDocProvider(stateDB, genDocProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	// Modify genesis chain ID and save to genesis file.
```

Comment:
We should use an app genesis here, otherwise we may overwrite the app genesis into a comet genesis

Reply:
@julienrbrt I think this addresses your concern, if you could look it over before merge that would be much appreciated!

https://github.com/cosmos/cosmos-sdk/pull/19280/commits/25bee82a0e7976bf4f861d7ba67965482dbae91d

Reply:
Yes, perfect, thank you!